Add benchmark for torch.distributed.pipeline.sync.Pipe

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49577 Add benchmark for torch.distributed.pipeline.sync.Pipe**
* #48638 Improve documentation for pipeline parallelism.
* #48630 Cleanup APIs for pipeline parallelism.
* #48470 Test pipeline parallelism works with DDP.

Repurposing the benchmarking from
https://github.com/facebookresearch/fairscale/blob/master/benchmarks/pipe.py
and pulling in a stripped down version of the benchmark into PyTorch.

Sample output:
```
Running benchmark with args: Namespace(batch_size=8, checkpoint='never', chunks=4, host='localhost', max_batch=10, num_decoder_layers=10, num_devices=4)
Number of parameters for model: 292833040
| batch     1 | wps 3593.07 | loss 25.98 | ppl 192556591553.37
| batch     2 | wps 4405.16 | loss 19.36 | ppl 256201548.33
| batch     3 | wps 4404.98 | loss 23.56 | ppl 17111244076.37
| batch     4 | wps 4413.25 | loss 27.11 | ppl 594561327825.83
| batch     5 | wps 4408.53 | loss 25.92 | ppl 181277705101.33
| batch     6 | wps 4385.64 | loss 24.92 | ppl 66592883598.50
| batch     7 | wps 4434.11 | loss 24.75 | ppl 56113635884.68
| batch     8 | wps 4441.25 | loss 24.88 | ppl 63666024212.82
| batch     9 | wps 4425.49 | loss 25.35 | ppl 101959669008.98
| batch    10 | wps 4421.05 | loss 25.34 | ppl 101597621863.94
Peak memory usage for GPUs: cuda:0: 2.38GiB, cuda:1: 3.04GiB, cuda:2: 3.04GiB, cuda:3: 3.67GiB,
```

Differential Revision: [D25628721](https://our.internmc.facebook.com/intern/diff/D25628721/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25628721/)!

pritamdamania committed:
Add benchmakr for torch.distributed.pipeline.sync.Pipe

Repurposing the benchmarking from
https://github.com/facebookresearch/fairscale/blob/master/benchmarks/pipe.py
and pulling in a stripped down version of the benchmark into PyTorch.

Sample output:
```
Running benchmark with args: Namespace(batch_size=8, checkpoint='never', chunks=4, host='localhost', max_batch=10, num_decoder_layers=10, num_devices=4)
Number of parameters for model: 292833040
| batch     1 | wps 3593.07 | loss 25.98 | ppl 192556591553.37
| batch     2 | wps 4405.16 | loss 19.36 | ppl 256201548.33
| batch     3 | wps 4404.98 | loss 23.56 | ppl 17111244076.37
| batch     4 | wps 4413.25 | loss 27.11 | ppl 594561327825.83
| batch     5 | wps 4408.53 | loss 25.92 | ppl 181277705101.33
| batch     6 | wps 4385.64 | loss 24.92 | ppl 66592883598.50
| batch     7 | wps 4434.11 | loss 24.75 | ppl 56113635884.68
| batch     8 | wps 4441.25 | loss 24.88 | ppl 63666024212.82
| batch     9 | wps 4425.49 | loss 25.35 | ppl 101959669008.98
| batch    10 | wps 4421.05 | loss 25.34 | ppl 101597621863.94
Peak memory usage for GPUs: cuda:0: 2.38GiB, cuda:1: 3.04GiB, cuda:2: 3.04GiB, cuda:3: 3.67GiB,
```

Differential Revision: [D25628721](https://our.internmc.facebook.com/intern/diff/D25628721/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25628721/)!

[ghstack-poisoned]

rohan-varma commented on benchmarks/distributed/pipeline/pipe.py:
```diff
@@ -0,0 +1,284 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
```
Nit: Not sure whats the guidelines around these copyrights, but it seems like only the pipeline files have this, and no other files in OSS pytorch (i.e. caffe2/torch directory)

pritamdamania87 replied:
I think we probably need it only for original torchgpipe files, will remove it for this.

rohan-varma commented on benchmarks/distributed/pipeline/pipe.py:
```diff
@@ -0,0 +1,284 @@
+
+    num_params = reduce(operator.add, (reduce(operator.mul, x.size()) for x in model.parameters()))
+    logging.info(f"training model, #prams = {num_params}")
+    vocab_size = 10000  # FIXME
```
Nit: What needs to be fixed here?

rohan-varma commented on benchmarks/distributed/pipeline/pipe.py:
```diff
@@ -0,0 +1,284 @@
+    import operator
+
+    num_params = reduce(operator.add, (reduce(operator.mul, x.size()) for x in model.parameters()))
+    logging.info(f"training model, #prams = {num_params}")
```
Nit: misspelling

rohan-varma commented on benchmarks/distributed/pipeline/pipe.py:
```diff
@@ -0,0 +1,284 @@
+            cur_loss = total_loss / log_interval
+            elapsed = time.time() - start_time
+            print(
+                "| batch {:5d} | wps {:5.2f} | loss {:5.2f} | ppl {:8.2f}".format(
```
Curious, What's the interpretation of pps here?

rohan-varma commented on benchmarks/distributed/pipeline/pipe.py:
```diff
@@ -0,0 +1,284 @@
+    os.environ.update({"MASTER_ADDR" : args.host})
+    os.environ.update({"MASTER_PORT" : "10638"})
+
+    rpc.init_rpc(
```
Why do we need to initialize RPC with a world_size of 1 if there's no RPC/cross host communication being used here?

rohan-varma reviewed (approved):
LGTM, although I'm curious why we need to init RPC here

pritamdamania committed:
Update on "Add benchmark for torch.distributed.pipeline.sync.Pipe"


Repurposing the benchmarking from
https://github.com/facebookresearch/fairscale/blob/master/benchmarks/pipe.py
and pulling in a stripped down version of the benchmark into PyTorch.

Sample output:
```
Running benchmark with args: Namespace(batch_size=8, checkpoint='never', chunks=4, host='localhost', max_batch=10, num_decoder_layers=10, num_devices=4)
Number of parameters for model: 292833040
| batch     1 | wps 3593.07 | loss 25.98 | ppl 192556591553.37
| batch     2 | wps 4405.16 | loss 19.36 | ppl 256201548.33
| batch     3 | wps 4404.98 | loss 23.56 | ppl 17111244076.37
| batch     4 | wps 4413.25 | loss 27.11 | ppl 594561327825.83
| batch     5 | wps 4408.53 | loss 25.92 | ppl 181277705101.33
| batch     6 | wps 4385.64 | loss 24.92 | ppl 66592883598.50
| batch     7 | wps 4434.11 | loss 24.75 | ppl 56113635884.68
| batch     8 | wps 4441.25 | loss 24.88 | ppl 63666024212.82
| batch     9 | wps 4425.49 | loss 25.35 | ppl 101959669008.98
| batch    10 | wps 4421.05 | loss 25.34 | ppl 101597621863.94
Peak memory usage for GPUs: cuda:0: 2.38GiB, cuda:1: 3.04GiB, cuda:2: 3.04GiB, cuda:3: 3.67GiB,
```

Differential Revision: [D25628721](https://our.internmc.facebook.com/intern/diff/D25628721/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D25628721/)!

[ghstack-poisoned]
